models/issue_comment.go

```diff
@@ -566,6 +601,153 @@ func CreateRefComment(doer *User, repo *Repository, issue *Issue, content, commi
 	return err
 }
 
+// ClearPullPushComment clear all the push commit on issue since fore push
+func ClearPullPushComment(issue *Issue) error {
+	_, err := x.Delete(&Comment{
+		Type:    CommentTypePullPushCommit,
+		IssueID: issue.ID,
+	})
+	return err
+}
+
+func createPullPushCommentsNonForcePush(doer *User, issues []*Issue, commits []*PushCommit) error {
+	sess := x.NewSession()
+	defer sess.Close()
+
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	for _, issue := range issues {
+		for i := len(commits) - 1; i >= 0; i-- {
+			_, err := createComment(sess, &CreateCommentOptions{
+				Type:         CommentTypePullPushCommit,
+				Doer:         doer,
+				Repo:         issue.Repo,
+				Issue:        issue,
+				CommitSHA:    commits[i].Sha1,
+				Content:      commits[i].Message,
+				RepoFullName: issue.Repo.FullName(),
+			})
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return sess.Commit()
+}
+
+// CreatePullPushComments creates commit comments when push commits to a pull request.
+func CreatePullPushComments(doer *User, issues []*Issue, commits []*PushCommit, isForcePush bool) error {
```

Comment:
As I mentioned in [a previous comment](https://github.com/go-gitea/gitea/pull/2548#discussion_r147324379), we can determine which comments need to be deleted from the `git rev-list` command that is run in [CommitRepoAction(..)](https://github.com/go-gitea/gitea/pull/2548/files#diff-7603238519a500de37f3e4b25cef1800R645). I believe that would be simpler and more efficient (since we have to run `git rev-list` regardless) than using the database.

Reply:
I have tested some times. It seems `git rev-list` returned different commits from `opts.Commits.Commits`.

Reply:
I have also done several examples (e.g. just removing commits, removing commits and adding new ones), and in every case the commit shas returned by `git rev-list` were exactly the commits whose comments would need to be removed (i.e. the commit that previously existed but do not exist anymore)

Do you have an example where `git rev-list` does not work?

Reply:
Yes. `git rev-list` is easier than current method. But it may add the master commits to this PR. When I rebase and push to a feature test. Some of commits are already in master and another not. We only add commit comments that not in target branch. See below screenshots is what the method you recommend. `13fabd0` has been pushed to `master` branch. Then I rebase `master` to `test` and then push to `777:test` it has been pull requested. Then commit comment `13fabd0` occupied but it should not.

![image](https://user-images.githubusercontent.com/81045/32478312-063ca650-c3bf-11e7-8b3d-4186ac86b970.png)

![image](https://user-images.githubusercontent.com/81045/32478307-f986d0c0-c3be-11e7-98c1-39df0039981e.png)

Reply:
@lunny I think he meant to use rev-list to detect commits that need to be deleted not added

Reply:
Yes, I see. but it's not enough. For my example, `3fabd0261` should not be as a commit comment, because it has been submited to maser. rev-list will not remove `3fabd0261`. So if force push, we have to find the HEAD of target branch and remove all commits before that.